### DISCUSSION TITLE
Issues with opencascade union operations

### DISCUSSION THREAD

**Discussion by timbod7** [donalffons/opencascade.js#89] [Q&A]
The opencascade.js project is impressive!

There's quite a few layers involved however, and it's hard to know how to debug issues that arise. An an example, here's a problem I don't understand in a model I built in cascadestudio:

https://github.com/zalo/CascadeStudio/issues/115

This could be a problem in

* cascadestudio's usage of opencascade.js
* opencascade.js implementation of opencascade
* opencascade itself

I've subsequently seen similar issues with union operations on other models, so presumably the errors are not that unusual. Is there anything I can do to get more informative errors messages?

**Comment by donalffons** [comment]
Hey, it's great to have more people checking the project out.

Debugging with OpenCascade.js is currently not so easy. However, I will investigate the topic a bit more in detail and will probably post some updates here, soon.

Here are some hints:
- Some errors will arise when C++ code throws an exception. Exceptions can have a significant performance and size impact. That's why currently, exception support is disabled in OpenCascade.js. Last time I checked, the performance impact seemed to be extremely large (I experienced a 5 times increase in loading a STEP file). However, I recently made some other tests (different project with different build settings) and experienced almost no performance impact. So, maybe there is a set of build settings that will make exceptions work nicely. Also, "native" WASM exceptions with smaller performance and size impacts will arrive in browsers probably soon (it's already in origin trial in Chrome). As of right now, enabling exception support in OpenCascade.js would require you to fork the library, i.e. this is not something that could be enabled with a "custom build".
- The error `___cxa_is_pointer_type is not defined` reported by you seems to be more of an Emscripten / build configuration issue (and not caused by C++ code throwing an exception). That error message has been reported a few times over at the Emscripten repository. Debugging these things can be hard and will most likely require a special debug build of OpenCascade.js
- CascadeStudio is using a rather old fork of OpenCascade.js which relies on an old version of Emscripten. Since then, there have been several improvements but also breaking changes. @zalo seemed to be open to upgrade to the newest version. But I assume that I would first have to release the next major version (i.e. non-beta) of OpenCascade.js, before they would even consider upgrading. So, I think the ball is in my field, here...
- Finally, there would be the option to enable debugging and assertions in Emscripten. Then you would probably get a stack trace including function names etc. for your errors. However, these things come with drastic file size and performance penalties and would require you to fork the library.

To sum this up: I think it should be possible to support Exceptions without too much effort. I have to finish a few things and finally release the next major version which would hopefully (among other things) allow an update to CascadeStudio. Distributing debug builds is not something I am planning to do as of right now.

**Reply by zalo** [reply]
[Doesn't look like the new OC.js helps with this particular kernel error...](https://cascade-studio-git-feat-embind-zalo.vercel.app/?code=pVRbb9owFH7nV3hvSQmQhNFdqj1Ug2qTNmlinfaAUOQmBxLVdZBtWsq0%2F75jOwlOQreHRYDJuX7fuZiBInz%2FkDDIJPlApuHVgKEIXxO5o2nBtyiN3lipEgXlWwaOyjW8IF%2BpysdpKb1pWL99%2B0wmJHob%2BjbEE2UsUXmR3nOQOmM0nlUpi22u7spDksFO5RpLpSg4B5FkBX0ABQIVDd6LPqJJndX6lnvV9u0EG3bSVlDoMxrloFXo0wniEnOgGKGi3Lt8iXtWbDZ7iaHKRxA54u4UIu6YnalV2DERNCv2MtnSHarDcWTVKd0lmEI1iVA5O6nyUhxd1bTqbrlLtFoT7bMekhhpeecyD1%2BEPOznq2pxRyUkNoWNhBlbACYk7lk2HYnDejQkqOgUoR911EfQcm1C9mrmmMWnDJ35MSgHkwm5JjhDqCXpMyt4JoqUMpKWpcgKThVIMtjseaqKEi1SNPUOUUAOcUAonhTPI57H2Ce%2FBgSftORSEQGphsbhiXy%2FB5Xm3kq7HaO1b6z0M%2F5ScLgtK018VhP%2FVdOKtuCZp8QeHNENTcHDtukXAWovOFmWCkl5qzAg%2BInWlsYSHkv2CF6tjYw2RO07PDUZX3PFnmhrx3lDmYTq8P3B78GpWAy2Xl0U3Y%2BmX6%2FDK0d4UILaIa9l5lqoRqWR1XaucFMwPCK9Xig1Yuxn%2B7mezxdz3K8wIp8Wy8Ur8uh0CbupvW1bW%2FMStAc0ICPPwhpWUHw9QAE5Lx1ZrkFFz2%2BnjDHljYG%2ByLYgvaZf%2F4FjZKXt7KEfNLGrWmHzLtdG6LtTQeZ4DYAAjvOii4JmGui63VG9o62WyhwYQzYfzeaA8HpbHPQvAf%2FUQMvNDdBh21r1rmPcc4z%2FmbXi69A1HJBvU6lbQblkzpL0r7FRG9i6BurU2%2BIxr7aKOJk%2FS3Ev3%2BM%2F%2FWKL2ZebtTHiJexKoSQBIUqBMPb6WtFkJaEoPexwLSEzfj849sdb2ZiBCbFuB5F47dlIuF8ZLmzBJF7bVUQcTqG3nD3r0AVHrhJM0ytUnx2RtxpUa2YHdqrvghB%2FRvqf%2Fo5m%2BDPT41cZnsVntGv%2FDw%3D%3D&gui=q1ZKzs8tyM9LzSvxS8xNVbJSSk4sTk5MSQ3LTC1X0lHyTS3OCEotVrIy0DOE84IS89KBSqMN9AwMdYxidZScE5MzUu2VrEqKSlNrAQ%3D%3D)

[There's a workaround that I posted in the other thread](https://github.com/zalo/CascadeStudio/issues/115), but I think it's just OpenCascade being weird and not any particular implementation detail...

**Reply by donalffons** [reply]
Thanks! I think I want to look at this error a bit closer at some point. I'm curious if OpenCascade itself gives more information on what's going wrong and why with some kind of exception message...

**Comment by timbod7** [comment]
Thanks for the reply.

cascadestudio (via hackaday) was my entrypoint to this space. But I'm keen to build some tooling of my own.

Next time I see an issue like this, I might try porting the code to a bare example in opencascadejs, so I can look at debugging in more detail.

**Reply by donalffons** [reply]
If you encounter any obstacles and/or need help with anything, please reach out. I'll try to help where I can.

Quick tip upfront: If you want to render anything in the 3D viewport, you will need to extract tessellation data from OpenCascade. You might want to look at the examples or at CascadeStudio for inspiration. The examples do rely on an now outdated ThreeJS feature (Geometry), which is why I would suggest not to use that - at least not long term. I think the easiest way of extracting data from OpenCascade for rendering (not just with ThreeJS) is to use a GLTF/GLB export. Here is a snippet:
```js
// create and initialize threejs and a scene available as three.scene
// ...
// create opencascade app
const app = new oc.TDocStd_Application();
oc.BinXCAFDrivers.DefineFormat(new oc.Handle_TDocStd_Application_2(app)); // needed for "BinXCAF" - type documents (i.e. with color, product structure, ...)

const visualizeShape = (s: TopoDS_Shape) => {
	// create doc, add shape label...
	const docHandle = new oc.Handle_TDocStd_Document_1();
	app.NewDocument(new oc.TCollection_ExtendedString_2("BinXCAF", true), docHandle);
	const doc = docHandle.get();
	const st = oc.XCAFDoc_DocumentTool.ShapeTool(doc.Main()).get();
	const shapeLabel = st.NewShape();
	st.SetShape(shapeLabel, s);

	// triangulate and export
	new oc.BRepMesh_IncrementalMesh_2(s, 0.1, false, 0.1, false);
	const cafWriter = new oc.RWGltf_CafWriter(new oc.TCollection_AsciiString_2("./file.glb"), true);
	cafWriter.Perform_2(docHandle, new oc.TColStd_IndexedDataMapOfStringString_1(), new oc.Message_ProgressRange_1())

	const file = oc.FS.readFile("./file.glb");
	app.SaveAs_1(docHandle, new oc.TCollection_ExtendedString_2("./export.xbf", false), new oc.Message_ProgressRange_1());

	new THREE_EX.GLTFLoader().load(
		URL.createObjectURL(new Blob([file])),
		gltf => {
			gltf.scene.traverse(o => {
				if((o as any).material?.metalness !== undefined) {
					(o as any).material.metalness = 0;
				}
			});
			three.scene.add(gltf.scene);
		}
	);
};
```
I think I also want to update OpenCascade.js' examples to use this pattern instead.

**Reply by weepingtown** [reply]
hi, @donalffons, 
asking for some help.
I'm trying to implement a program that converts the step file to gltf format. 
I have read your example program, and found that it directly reads the .step file using ```STEPControl_Reader```, which will result in the loss of component information. As recommended by opencascade, it should be replaced by ```STEPCAFControl_Reader```. 
I have a sample code for a c++ implementation([step2gltf](https://draw.sview.ru/xde-export-step2gltf/)), which works well, and now I need to rewrite it to the JavaScript version.
I mainly use C/C++ in my work and am not familiar with JavaScript. When using opencascade.js, I have encountered some difficulties, such as the following code, and I do not know how to rewrite it to JavaScript, especially the implementation of iterators.
```c++
TopoDS_Compound aCompound;
BRep_Builder    aBuildTool;
aBuildTool.MakeCompound (aCompound);
for (TDF_LabelSequence::Iterator aRootIter (aRootLabels); aRootIter.More(); aRootIter.Next())
{
  const TDF_Label& aRootLabel = aRootIter.Value();
  TopoDS_Shape aRootShape;
  if (XCAFDoc_ShapeTool::GetShape (aRootLabel, aRootShape))
  {
    aBuildTool.Add (aCompound, aRootShape);
  }
}
```

I would appreciate it if you could help with some guidance.

**Reply by zalo** [reply]
If you're talking about the iterator syntax, [there are a few examples of related things in CascadeStudio](https://github.com/zalo/CascadeStudio/blob/master/js/CADWorker/CascadeStudioStandardLibrary.js#L203-L210).

The aspect where I call `oc.TopoDS.prototype.Solid()` on the current element is just to get a version of that solid that can be used outside the loop (may not be necessary for adding to the BRep_Builder).

**Reply by donalffons** [reply]
I assume `aRootLabels` is of type `TDF_LabelSequence`?! In this case, the following might work (untested):
```js
const aBuildTool = new oc.BRep_Builder();
aBuildTool.MakeCompound (aCompound);
for(let i = 1; i <= aRootLabels.Length(); i++) {
  const aRootLabel = aRootLabels.Value(i);
  const shape = oc.XCAFDoc_ShapeTool.GetShape_2(label);
  if(!shape.isNull()) {
    aBuildTool.Add(aCompound, shape);
  }
}
```

The `TDF_LabelSequence::Iterator` syntax is currently not supported directly by OpenCascade.js but if you need to access it, there might be a way using`additionalCppCode` (e.g. a `typedef` or some other construct).

**Reply by weepingtown** [reply]
It seems to work. I will test it next week. Thank you very much for your reply.

**Reply by weepingtown** [reply]
> If you're talking about the iterator syntax, [there are a few examples of related things in CascadeStudio](https://github.com/zalo/CascadeStudio/blob/master/js/CADWorker/CascadeStudioStandardLibrary.js#L203-L210).
> 
> The aspect where I call `oc.TopoDS.prototype.Solid()` on the current element is just to get a version of that solid that can be used outside the loop (may not be necessary for adding to the BRep_Builder).

 
Very helpful.
In fact, what I'm doing is reading a step file and rendering it in frontend with three.js. Cascadestudio provides rich examples, and I'll be reading your work carefully over the next few days. Thank you so much for your great job.

**Reply by weepingtown** [reply]
hi,
Just now, I verified your suggestions, and they all work properly. Thank you very much, @zalo @donalffons .
As zalo says, iterators and BRep_Builder are not really necessary for me. I actually just need  ```aRootLabels.First()```.

During the test, I found an issue. When a node without triangulation data exists in the .stp file, the RWGltf_CafWriter will skip this node. However, in the generated .gltf file, the mesh index of this node is not updated accordingly. As a result, the gltf file cannot be parsed properly.
I used Open Cascade 7.6 to create a C++ demo program for verification and found that the problem did not recur.So I guess it's a problem specific to opencascade.js. I'll submit an issue later to explain this in detail.

**Comment by donalffons** [comment]
Just another remark about exceptions: I recently made some prototyping with "native" WASM exceptions (which is [behind a flag in Chrome](https://v8.dev/blog/v8-release-95#exception-handling) since a few days). I experienced a file size increase of 24% for a custom build of the library. Also, chrome crashed during most of my testing (v8 issue is [here](https://bugs.chromium.org/p/v8/issues/detail?id=12255)). When it did not crash, it seems like the performance impact was small.

That makes me very hopeful that we will have proper support for exceptions at some point in the not too distant future. However, I think we would either have to wait for other engines (firefox, node, safari) to catch up or distribute OpenCacade.js in different "flavors" (e.g. LTS, dev, multi/single-threaded, Wasm64-enabled at some point in the future, ...?). Not sure, but that's probably more of a long-term problem...
